pkg/ruler/base/ruler.go

```diff
@@ -642,6 +704,77 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring r
 	return result
 }
 
+// filterRules returns a list of rule groups, each with a single rule, ONLY for rules that are "owned" by the given ring instance.
+func filterRules(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring ring.ReadRing, instanceAddr string, logger log.Logger, ringCheckErrors prometheus.Counter) []*rulespb.RuleGroupDesc {
+	// Return one rule group per rule that is owned by this ring instance.
+	// One rule group is returned per rule because Prometheus executed rule groups concurrently but rules within a group sequentially;
+	// we are sharding by rule here to explicitly *avoid* sequential execution since Loki does not need this.
+	var result = make([]*rulespb.RuleGroupDesc, 0, len(ruleGroups))
+
+	for _, g := range ruleGroups {
+		logger = log.With(logger, "user", userID, "namespace", g.Namespace, "group", g.Name)
+
+		for _, r := range g.Rules {
+			logger = log.With(logger, "rule", getRuleIdentifier(r))
+
+			owned, err := instanceOwnsRule(ring, g, r, instanceAddr)
+			if err != nil {
+				ringCheckErrors.Inc()
+				level.Error(logger).Log("msg", "failed to check if the ruler replica owns the rule", "err", err)
+				continue
+			}
+
+			if !owned {
+				level.Debug(logger).Log("msg", "rule not owned, ignoring")
+				continue
+			}
+
+			level.Debug(logger).Log("msg", "rule owned")
+
+			// clone the group and replace the rules
+			clone := cloneGroupWithRule(g, r)
+			if clone == nil {
```

Comment:
That would make the tests much uglier because we'd have to check for multiple return values; I agree with you but I think the change would add little actual value